crates/subspace-networking/src/shared.rs

```diff
@@ -79,3 +85,61 @@ impl Shared {
         }
     }
 }
+
+// Supports exact keys for Kademlia. It uses the Multihash as the canonical type
+// and reduces all the conversions to Multihash conversion.
+// It supports only Sr25519 key pairs for PeerId and Identity multihashes.
+#[derive(Clone, Debug)]
+pub struct IdendityHash(GenericArray<u8, U32>);
+
+impl PreimageIntoKeyBytes<Multihash> for IdendityHash {
+    fn preimage_into_key_bytes(multihash: &Multihash) -> KeyBytes<Self> {
+        const SHA256_HASH_SIZE: usize = 32;
+
+        // Identity hasher for 32-bytes digest
+        if multihash.code() == u64::from(Code::Identity)
+            && multihash.digest().len() == SHA256_HASH_SIZE
+        {
+            let array = GenericArray::from_slice(multihash.digest());
+            return KeyBytes::from_unchecked(*array);
+        }
+
+        panic!(
+            "Unsupported multihash type. Expected Identity:{:?}",
+            multihash
+        )
+    }
+}
+
+impl PreimageIntoKeyBytes<PeerId> for IdendityHash {
+    fn preimage_into_key_bytes(peer_id: &PeerId) -> KeyBytes<Self> {
+        let multihash: Multihash = if let Ok(public_key) =
+            identity::PublicKey::from_protobuf_encoding(peer_id.as_ref().digest())
+        {
+            if let identity::PublicKey::Sr25519(pk) = public_key {
+                Code::Identity.digest(&pk.encode())
+            } else {
+                panic!("Unsupported public key type. Expected Sr25519")
+            }
+        } else {
+            // No protobuf encoding: it's PeerId::random()
+            *peer_id.as_ref()
```

Comment:
```suggestion
            peer_id.into()
```

Reply:
It's not an identical expression (doesn't compile this way).

Reply:
Hm... I see there is `impl AsRef<Multihash> for PeerId` you're dereferencing afterwards, but there is also `impl From<PeerId> for Multihash`. The reason it doesn't compile is `&PeerId`, but `(*peer_id).into()` does and clearer shows what you want to do.